src/profiler.ts

```diff
@@ -13,57 +13,203 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+
+import * as delay from 'delay';
+import * as http from 'http';
 import * as path from 'path';
+import * as zlib from 'zlib';
 
-import {AuthenticationConfig, Common, Logger, Service, ServiceConfig} from '../third_party/types/common-types';
+import {perftools} from '../src/profile';
+import {AuthenticationConfig, Common, Logger, Service, ServiceConfig, ServiceObject, ServiceObjectConfig} from '../third_party/types/common-types';
 
-import {Config} from './config';
+import {ProfilerConfig} from './config';
 import {HeapProfiler} from './profilers/heap-profiler';
 import {TimeProfiler} from './profilers/time-profiler';
 
-const pjson = require('../../package.json');
 const common: Common = require('@google-cloud/common');
+const pjson = require('../../package.json');
+const API = 'https://cloudprofiler.googleapis.com/v2';
 
-export interface ProfilerConfig extends AuthenticationConfig {
-  projectId: string;
-  logLevel: number;
-  serviceContext: {service: string; version?: string;};
-  instance: string;
-  zone: string;
-  disableTime: boolean;
-  disableHeap: boolean;
+// Interface for body of response from Stackdriver Profiler API when creating
+// profile and used as body of request to Stackdriver Profiler API when
+// uploading a profile.
+// Public for testing.
+export interface RequestProfile {
+  name?: string;
+  profileType?: string;
+  duration?: any;
+  profileBytes?: string;
+  labels?: {instance?: string; zone?: string};
 }
 
-// TODO: finish implementing Profiler.
-// TODO: add stop() method to stop profiling.
+// Converts a profile to a compressed, base64 encoded string.
+function profileBytes(p: perftools.profiles.IProfile): Promise<string> {
+  let time = new Date((p.timeNanos as number) / 1000000);
+  let pwriter = perftools.profiles.Profile.encode(p);
+  let buffer = new Buffer(pwriter.finish());
+  return new Promise<string>((resolve, reject) => {
+    zlib.gzip(buffer, (err: Error, gzBuf: Buffer) => {
```

Comment:
It would be a lot cleaner if you used `util.promisify` or `pify` to promisify this API and then simply awaited `zlib.gzip`. You'll need to make this function async.

There are a few other examples in this PR that would benefit from the same pattern.

Reply:
I was able to use pify here, but I couldn't figure out how to get this to work for serviceObject.request()  -- I got errors at run time related to serviceObject being undefined. Could you point me to any examples for promisifying a method?